tools/vscode_sql_extraction/src/highlighter.ts

```diff
@@ -0,0 +1,100 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+import {Query, locationToRange} from './query';
+import randomColor from 'randomcolor';
+
+/**
+ * Highlights found queries in the text editor.
+ */
+export class Highlighter {
+  decorations: vscode.TextEditorDecorationType[] = [];
+
+  /**
+   * Highlights all queries in the open document.
+   *
+   * @param openEditor currently open editor.
+   * @param workspaceRoot root path of the open workspace.
+   * @param queries list of found queries.
+   */
+  highlight(
+    openEditor: vscode.TextEditor,
+    workspaceRoot: string,
+    queries: Query[]
+  ) {
+    const currentQueries = this.getAllQueriesInCurrentFile(
+      openEditor,
+      workspaceRoot,
+      queries
+    );
+    if (currentQueries.length <= 0) {
+      return;
+    }
+
+    currentQueries.forEach((query, index) => {
+      const ranges: vscode.Range[] = [];
+
+      const workRemaining = [query.query];
+      while (workRemaining.length > 0) {
+        const fragment = workRemaining.pop()!;
+        if (fragment.literal) {
+          ranges.push(locationToRange(fragment.location));
+        } else {
+          // recurse for child fragments
+          fragment.complex!.forEach(child => {
+            workRemaining.push(child);
+          });
+        }
+      }
+
+      query.usages.forEach(usage => {
+        ranges.push(locationToRange(usage));
+      });
+
+      openEditor.setDecorations(this.getColor(index), ranges);
+    });
+  }
+
+  /**
+   * Gets all queries located in the currently open file.
+   *
+   * @param openEditor currently open editor.
+   * @param workspaceRoot root path of the open workspace.
+   * @param queries list of found queries.
+   */
+  private getAllQueriesInCurrentFile(
+    openEditor: vscode.TextEditor,
+    workspaceRoot: string,
+    queries: Query[]
+  ) {
+    let openPath = openEditor.document.uri.path.toString();
+    if (!path.isAbsolute(openPath)) {
+      openPath = path.join(workspaceRoot, openPath);
+    }
+
+    return queries.filter(query => {
+      let filePath = query.file;
+      if (!path.isAbsolute(filePath)) {
+        filePath = path.join(workspaceRoot!, query.file);
+      }
+      return filePath === openPath;
+    });
+  }
+
+  /**
+   * Updates the cache if needed, and then returns the relevant decoration.
+   *
+   * @param index decoration number to return.
+   */
+  private getColor(index: number): vscode.TextEditorDecorationType {
```

Comment:
In another pull request, please change getColor() to use the highlight color of the current theme.

Reply:
Good idea! Currently, I'm highlighting each query using a different color, but I can just have a few theme colors and cycle between them.